File: diagflow/core/memory.py
"""
Memory and evidence pooling for multi-agent diagnosis.

Two complementary constructs:
  1. SessionMemory — keeps conversation context within a single diagnosis,
     managing the sliding window of messages sent to the LLM.
  2. EvidencePool — a shared, structured store where specialist agents
     deposit findings. The orchestrator reads from this pool when
     synthesising the final diagnosis. This decouples agents — they never
     talk to each other directly, only through evidence.
"""

from __future__ import annotations

import json
import logging
import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Evidence
# ---------------------------------------------------------------------------

@dataclass
class Evidence:
    """A single finding from one diagnostic step."""
    source_agent: str            # e.g. "log_analyst", "metrics_analyst"
    category: str                # e.g. "log_error", "config_anomaly", "metric_spike"
    summary: str                 # human-readable finding
    detail: str                  # full context (e.g. matched log lines)
    confidence: float = 0.5      # 0.0 – 1.0
    timestamp: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    raw_data: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return {
            "source_agent": self.source_agent,
            "category": self.category,
            "summary": self.summary,
            "detail": self.detail,
            "confidence": self.confidence,
            "timestamp": self.timestamp,
        }

    def short_str(self) -> str:
        return f"[{self.source_agent}] {self.summary} (conf={self.confidence:.1f})"


# ---------------------------------------------------------------------------
# Evidence Pool
# ---------------------------------------------------------------------------

class EvidencePool:
    """Thread-safe collection of evidence gathered during diagnosis.

    Agents write into the pool; the orchestrator reads from it.
    This is the backbone of our multi-agent architecture.
    """

    def __init__(self) -> None:
        self._items: list[Evidence] = []
        self._lock = threading.Lock()

    def add(self, evidence: Evidence) -> None:
        with self._lock:
            self._items.append(evidence)

    def add_many(self, evidences: list[Evidence]) -> None:
        with self._lock:
            self._items.extend(evidences)

    def all(self) -> list[Evidence]:
        with self._lock:
            return list(self._items)

    def by_agent(self, agent_name: str) -> list[Evidence]:
        with self._lock:
            return [e for e in self._items if e.source_agent == agent_name]

    def by_category(self, category: str) -> list[Evidence]:
        with self._lock:
            return [e for e in self._items if e.category == category]

    def high_confidence(self, threshold: float = 0.7) -> list[Evidence]:
        with self._lock:
            return [e for e in self._items if e.confidence >= threshold]

    def summary(self) -> str:
        """Compact summary for inclusion in the LLM prompt."""
        with self._lock:
            items = list(self._items)
        lines = ["=== Evidence Pool ==="]
        for e in items:
            lines.append(f"  {e.short_str()}")
        if not items:
            lines.append("  (no evidence collected yet)")
        return "\n".join(lines)

    def clear(self) -> None:
        with self._lock:
            self._items.clear()


# ---------------------------------------------------------------------------
# Session Memory
# ---------------------------------------------------------------------------

class SessionMemory:
    """Manages the conversation window for a single diagnosis session.

    Implements a simple sliding-window: once the estimated token count
    exceeds ``max_tokens``, earlier tool responses are summarised away
    to keep the context within bounds.
    """

    def __init__(self, max_tokens: int = 24_000) -> None:
        self.max_tokens = max_tokens
        self.messages: list[dict[str, Any]] = []

    def add_user(self, content: str) -> None:
        self.messages.append({"role": "user", "content": content})

    def add_assistant(self, content: str | None, tool_calls: list | None = None) -> None:
        msg: dict[str, Any] = {"role": "assistant", "content": content or ""}
        if tool_calls:
            msg["tool_calls"] = [
                {"id": tc.id, "type": "function",
                 "function": {"name": tc.name, "arguments": json.dumps(tc.arguments)}}
                for tc in tool_calls
            ]
        self.messages.append(msg)

    def add_tool_result(self, tool_call_id: str, content: str) -> None:
        self.messages.append({
            "role": "tool",
            "tool_call_id": tool_call_id,
            "content": content,
        })

    def get_messages(self) -> list[dict[str, Any]]:
        return self.messages

File: internal/memory/memory.go
// Package memory mirrors diagflow/core/memory.py — evidence pooling and
// session memory shared across the diagnostic engine.
package memory

import (
	"fmt"
	"sync"
	"time"
)

// Evidence is a single finding from one diagnostic step.
type Evidence struct {
	SourceAgent string         `json:"source_agent"`
	Category    string         `json:"category"`
	Summary     string         `json:"summary"`
	Detail      string         `json:"detail"`
	Confidence  float64        `json:"confidence"`
	Timestamp   string         `json:"timestamp"`
	RawData     map[string]any `json:"-"`
}

// NewEvidence builds an Evidence with a zero-time UTC timestamp default.
func NewEvidence(sourceAgent, category, summary, detail string, confidence float64) Evidence {
	return Evidence{
		SourceAgent: sourceAgent,
		Category:    category,
		Summary:     summary,
		Detail:      detail,
		Confidence:  confidence,
		Timestamp:   time.Now().UTC().Format(time.RFC3339),
	}
}

// ToDict returns a map representation (excluding RawData), matching
// Evidence.to_dict in Python.
func (e Evidence) ToDict() map[string]any {
	return map[string]any{
		"source_agent": e.SourceAgent,
		"category":     e.Category,
		"summary":      e.Summary,
		"detail":       e.Detail,
		"confidence":   e.Confidence,
		"timestamp":    e.Timestamp,
	}
}

// ShortStr renders "[source_agent] summary (conf=0.9)".
func (e Evidence) ShortStr() string {
	return fmt.Sprintf("[%s] %s (conf=%.1f)", e.SourceAgent, e.Summary, e.Confidence)
}

// EvidencePool is a thread-safe collection of evidence gathered during diagnosis.
type EvidencePool struct {
	mu    sync.Mutex
	items []Evidence
}

// Add appends one evidence item.
func (p *EvidencePool) Add(e Evidence) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.items = append(p.items, e)
}

// AddMany appends several items.
func (p *EvidencePool) AddMany(es []Evidence) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.items = append(p.items, es...)
}

// All returns a snapshot copy of the pool.
func (p *EvidencePool) All() []Evidence {
	p.mu.Lock()
	defer p.mu.Unlock()
	out := make([]Evidence, len(p.items))
	copy(out, p.items)
	return out
}

// ByAgent filters by source agent.
func (p *EvidencePool) ByAgent(agent string) []Evidence {
	p.mu.Lock()
	defer p.mu.Unlock()
	var out []Evidence
	for _, e := range p.items {
		if e.SourceAgent == agent {
			out = append(out, e)
		}
	}
	return out
}

// ByCategory filters by category.
func (p *EvidencePool) ByCategory(category string) []Evidence {
	p.mu.Lock()
	defer p.mu.Unlock()
	var out []Evidence
	for _, e := range p.items {
		if e.Category == category {
			out = append(out, e)
		}
	}
	return out
}

// HighConfidence filters by confidence >= threshold.
func (p *EvidencePool) HighConfidence(threshold float64) []Evidence {
	p.mu.Lock()
	defer p.mu.Unlock()
	var out []Evidence
	for _, e := range p.items {
		if e.Confidence >= threshold {
			out = append(out, e)
		}
	}
	return out
}

// Summary renders a compact multi-line summary for LLM prompts.
func (p *EvidencePool) Summary() string {
	items := p.All()
	lines := []string{"=== Evidence Pool ==="}
	for _, e := range items {
		lines = append(lines, "  "+e.ShortStr())
	}
	if len(items) == 0 {
		lines = append(lines, "  (no evidence collected yet)")
	}
	out := ""
	for i, l := range lines {
		if i > 0 {
			out += "\n"
		}
		out += l
	}
	return out
}

// Clear empties the pool.
func (p *EvidencePool) Clear() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.items = p.items[:0]
}

// Len returns the number of items.
func (p *EvidencePool) Len() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	return len(p.items)
}

// Message is a single session message (role + content).
type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// SessionMemory manages the conversation window for a diagnosis session.
type SessionMemory struct {
	MaxTokens int
	Messages  []Message
}

// NewSessionMemory returns a SessionMemory with the default 24k token window.
func NewSessionMemory() *SessionMemory {
	return &SessionMemory{MaxTokens: 24000}
}

// AddUser appends a user message.
func (s *SessionMemory) AddUser(content string) {
	s.Messages = append(s.Messages, Message{Role: "user", Content: content})
}

// AddAssistant appends an assistant message.
func (s *SessionMemory) AddAssistant(content string) {
	s.Messages = append(s.Messages, Message{Role: "assistant", Content: content})
}

// GetMessages returns the message slice.
func (s *SessionMemory) GetMessages() []Message {
	return s.Messages
}
